### ISSUE TITLE
SDL_ANDROID_BLOCK_ON_PAUSE should be removed, instead please fix SDL_WaitEvent

### ISSUE THREAD

**Issue by SDLBugzilla** [libsdl-org/SDL#3193]
# This bug report was migrated from our old Bugzilla tracker.

**Reported in version:** 2.0.9
**Reported for operating system, platform:** Android (All), x86_64

# Comments on the original bug report:

On 2019-03-07 12:45:08 +0000, Ellie wrote:

> SDL_ANDROID_BLOCK_ON_PAUSE somewhat looks like a hack around the fact that SDL_WaitEvent wastes battery due to not being implemented in a smart way. This hack breaks the semantics of SDL_PollEvent in surprising and not-cross-platform ways.
> 
> I suggest the following changes to fix this:
> 
> - remove SDL_ANDROID_BLOCK_ON_PAUSE / default it to 0
> - SDL_WaitEvent should be fixed at least on Android to actually block
> 
> Apps can then choose to use SDL_WaitEvent if in the background if they want this blocking behavior. This solves: 1. the weird unexpected API semantics difference, 2. that this cannot be chosen at runtime without recompiling SDL2, 3. that there's this weird duplication of functionality which SDL_WaitEvent already should provide

On 2019-04-02 07:20:50 +0000, Sylvain wrote:

> SDL_WaitEvent is actually doing SDL_PollEvent + SDL_Delay: 
> 
> https://hg.libsdl.org/SDL/file/d5b4d374a312/src/events/SDL_events.c#l697

On 2019-04-02 09:20:20 +0000, Ellie wrote:

> Yes and that makes SDL_WaitEvent somewhat useless in terms of battery saving, even though at a first glance it would be the obvious choice! So fixing it to be truly blocking would make it the obvious usable choice and make that on pause "feature" completely unnecessary

On 2019-04-17 02:53:42 +0000, Sam Lantinga wrote:

> SDL_WaitEvent() can't block because events come from multiple sources that can't be aggregated in a cross-platform way.

On 2019-04-17 07:00:33 +0000, Ellie wrote:

> Isn't the point of SDL2 to solve these issues between different platforms? If that currently doesn't work with various internal sources, then it sounds a little like that should be refactored.
> 
> Also it would already help a lot if at least it worked properly on mobile platforms, if getting this to work on all platforms is too much work.
> 
> I can tell you, I have lurked in freenode/irc and even familiar SDL2 users are always surprised that SDL_WaitEvent is close to a busy loop rather than a proper block-until-event. It's really not expected behavior, and it would IMHO in the long term make much more sense to see how to fix it rather than add hacks like SDL_ANDROID_BLOCK_ON_PAUSE to work around this - at least if we're talking purely from a sane API design angle.
> 
> Of course if this is way too complicated to be feasible then maybe it's not worth it. But I can't judge that, I'm not familiar with the affected code

On 2019-04-23 03:51:41 +0000, Sam Lantinga wrote:

> Yes, it is too complicated to be feasible, in my opinion. However, I'll leave this bug open in case someone wants to contribute a patch they feel is valuable.

**Comment by icculus** [comment]
I still think a fully-waitable SDL is possible in the most common cases, even if we have to fall into busy-looping to cover some scenarios, so I'm tossing this into the SDL3 milestone to consider.

**Comment by superfury** [comment]
The main issue why the define for SDL_ANDROID_BLOCK_ON_PAUSE is there is because this very issue. It's because a backgrounded app (think music player or the like, or even messaging apps, games that need to continue backgrounded etc.) need to continue running (can't block) in the background. And SDL_PollEvent itself is blocking in that case(when the define isn't used to override), preventing the app from properly running in the background (blocking the main loop) if said define isn't used (also affecting audio processing for audio players by blocking as well, and messing up game timings).

My app itself only uses SDL_PollEvent, but if that's blocking it can't run and time properly anymore (not to mention generated audio being muted), as one of the core message handling is blocking, which it should never be by default.

**Comment by slouken** [comment]
Yeah, we should revisit this for SDL3. @1bsyl, with SDL's lifecycle improvements, do we still need this?

**Comment by superfury** [comment]
In fact, thinking about it, it should be like this:
- Blocking on pause should be disabled by default (handled by the app itself).
- Audio muting should be handled by the app as well (context saving of audio (OpenGL) being handled by SDL to allow backgrounded playback of audio like when using the override).
- The DID/WILL events would indicate to the app if it should optionally mute audio or pause(block by delaying) itself fully, with full freedom, where the app itself calls SDL_Delay or mute audio samples generated. In fact, this is already the case?
- SDL_WaitEvent can and should delay if no events are queued, as that's it's entire purpose.
- SDL_PollEvent can't and shouldn't delay in any way. It's called polling instead of busywaiting for a reason.

**Comment by 1bsyl** [comment]
btw
there is SDL_HINT_ANDROID_BLOCK_ON_PAUSE_PAUSEAUDIO
so that it can still play audio in background (it simply doesn't pause the audio in bg).

Blocking on pause seems to me a good default for a game...

**Comment by superfury** [comment]
> btw there is SDL_HINT_ANDROID_BLOCK_ON_PAUSE_PAUSEAUDIO so that it can still play audio in background (it simply doesn't pause the audio in bg).
> 
> Blocking on pause seems to me a good default for a game...

Yeah, I know.
Both options combine into a bigger effect though? They are handled in the same functions as far as I remember (the Blocking and NonBlocking ones).

The main point is that them being Blocking or Muting (losing GL context for audio rendering actually) by default is just a compatibility option for older Android apps. None of the other OS builds of SDL2 do this, for examle.
In fact, doing that on other OSes (like Windows) would break apps on such platforms afaik (since they don't expect it to happen), again messing up timings and audio. Android is the only platform doing this from what I know.

As I said, it should be the app's own decision to mute audio (or continue rendering something in the backgrou nda) or pause itself (based on the DID/WILL events or window minimize status) if the developer wants to, not SDL forcing one or the other (mute, blocking, or both)? The decision and optional usage of the old way (blocking the thread calling and losing audio capability for rendering) is a compatibility case here. Perhaps default to the new way (NonBlocking behaviour and continue audio playback and letting the app mute by rendering silence samples itself if the dev intends to) for SDL3 would be the way? That would improve both cases, as apps will need to be changed for SDL3 anyways?

**Comment by slouken** [comment]
I agree with @superfury, let's move the logic for this into the application for SDL3.

**Comment by 1bsyl** [comment]
I think this may be confusing and people are going start reporting issues and/or event not seeing their issues.

but we can try and that would be  ?
```diff
src/video/android/SDL_androidvideo.c

-    block_on_pause = SDL_GetHintBoolean(SDL_HINT_ANDROID_BLOCK_ON_PAUSE, SDL_TRUE);
+    block_on_pause = SDL_GetHintBoolean(SDL_HINT_ANDROID_BLOCK_ON_PAUSE, SDL_FALSE);

-    videodata->pauseAudio = SDL_GetHintBoolean(SDL_HINT_ANDROID_BLOCK_ON_PAUSE_PAUSEAUDIO, SDL_TRUE);
+    videodata->pauseAudio = SDL_GetHintBoolean(SDL_HINT_ANDROID_BLOCK_ON_PAUSE_PAUSEAUDIO, SDL_FALSE);

```

Also, "pause audio thread" feature is not exported as a public API.
here, the thread is locked (`SDL_LockMutex(device->lock); `), not sure how different is from putting silence.

Turning the non-blocking into a blocking loop, from application point of view, isn't similar as putting a SDL_Delay(). the blocking loop is done with a SDL_WaitSemaphore(). // sem_wait().
this also makes sure it's synchronized with the java SurfaceView states.

**Comment by superfury** [comment]
Well, people shouldn't be getting issues if they follow the official SDL3 migration guides and documentation (as SDL3 isn't even officially released yet). They're starting with a fresh SDL3 project after all (or using the migration guide anyways to migrate, if done properly).

And ofc the migration guide and SDL3 functions' wiki being updated on these changes is required for that.

If people will need to migrate to SDL3, they would need to check for these kinds of things anyways (and should check the migration guide). The same for new SDL3 projects, as those are clean and won't be directly affected by this (assuming the documentation is up-to-date on this behaviour).

**Comment by superfury** [comment]
Any progress on this change (by @1bsyl 1bsyl)? I don't see the change yet in the prerelease of SDL 3.1.1?

**Comment by 1bsyl** [comment]
not from me, not sure what SDL3 behavior change is expected here ? 
I haven't re-read this, but from me last message it could be simply changing the default values of the 2 hints:
SDL_HINT_ANDROID_BLOCK_ON_PAUSE
SDL_HINT_ANDROID_BLOCK_ON_PAUSE_PAUSEAUDIO

to be "false" by default.

**Comment by superfury** [comment]
> not from me, not sure what SDL3 behavior change is expected here ? I haven't re-read this, but from me last message it could be simply changing the default values of the 2 hints: SDL_HINT_ANDROID_BLOCK_ON_PAUSE SDL_HINT_ANDROID_BLOCK_ON_PAUSE_PAUSEAUDIO
> 
> to be "false" by default.

Yes. That what it effectively boils down to the solution of the issue (of course with changing the documention wiki and header documention to reflect that, as well as the SDL3 migration guide noting the change).

**Comment by 1bsyl** [comment]
Not really sure with want this.
Most app usually doesn't play music in background
nor try to render in background or eat 100% in bg.
Currently, it's seamlessly working when you move an app from desktop to mobile. I wouldn't make the platform flimsy with this. 
One can set the hint if needed, in very rare occasion I guess.

**Comment by superfury** [comment]
> Not really sure with want this. Most app usually doesn't play music in background nor try to render in background or eat 100% in bg. Currently, it's seamlessly working when you move an app from desktop to mobile. I wouldn't make the platform flimsy with this. One can set the hint if needed, in very rare occasion I guess.

Not exactly seamlessly though. If on Windows you minimize the app (rough equivalent of Android behaviour) the app won't block the thread handling the events (thus everything the app does continues, just no SDL3 events happen). But Android otoh does block the calling thread, so compared to the other platforms it's the one that's behaving different?
So it's already Android vs other platforms that is incompatible, which is why this override exists in the first place (Windows for example doesn't block, so apps will merrily run when minimized. But Android apps will start to block the main thread when the app is ported over straight from Windows (probably Linux as well, don't know about other platforms)).

**Comment by slime73** [comment]
I think Android is closest to iOS – although it does have differences. On iOS if you minimize an app, the OS pauses the entire app aside from a few special cases for running small low-power things in the background. As a user of an iPhone this is pretty desirable, but I can't really speak for Android users.

**Comment by superfury** [comment]
You reacted before I could finish editing. Anyways...

> Not really sure with want this. Most app usually doesn't play music in background nor try to render in background or eat 100% in bg. Currently, it's seamlessly working when you move an app from desktop to mobile. I wouldn't make the platform flimsy with this. One can set the hint if needed, in very rare occasion I guess.

Not exactly seamlessly though. If on Windows you minimize the app (rough equivalent of Android behaviour) the app won't block the thread handling the events (thus everything the app does continues, just no SDL3 events happen). But Android otoh does block the calling thread, so compared to the other platforms it's the one that's behaving different?
So it's already Android vs other platforms that is incompatible, which is why this override exists in the first place (Windows for example doesn't block, so apps will merrily run when minimized. But Android apps will start to block the main thread when the app is ported over straight from Windows (probably Linux as well, don't know about other platforms)).
Basically it's Android behaviour (blocking) vs other platforms (except IOS perhaps) that don't block that's part of the main issue.
And like you said, changing it because of porting behaviour is incompatible no matter what default you pick (either you'd change behaviour on Android or you'd change behaviour on all other platforms). Although a sensible default would be the majority I'd think (thus all other platforms, with it not blocking)?

**Comment by slouken** [comment]
I've reviewed the blocking behavior on Android, and made it more closely match iOS. SDL_ANDROID_BLOCK_ON_PAUSE is still the default, but if you turn it off, you'll use low CPU in the background instead of spinning using high CPU.

**Comment by slouken** [comment]
The latest commit also makes it so waiting on events doesn't spin on Android.

**Comment by superfury** [comment]
> The latest commit also makes it so waiting on events doesn't spin on Android.

But does it allow audio rendering while the app is in the background? It's fine that the threads keep running, but in my app's case it also has a media player. It has an option inside it to enable playing audio backgrounded (or to mute it optionally when backgrounded). The audio thread is also required to clear buffers (shouldn't cause any problems I think, other than audio getting lost once the rendering buffers are full. If not, the app would hang waiting for more room, which causes a wait loop, but I'm pretty positive that wasn't blocking like that).

Anyways, my app can and does in fact keep rendering sound when backgrounded on Android (or any OS for that matter), just rendering zeroed samples when instructed to mute outputted samples. The same kind pf thing with recording devices detected (muting if to mute during backgrounded operation).
The main thread as it keeps (the x86 emulator) going just keeps rendering samples in realtime speeds (or as close possible) according to emulated hardware converted to SDL's reported samplerate (w/ low-pass filter before resampling).
The main app startup routine will set the SDL2/3 BLOCK_ON_PAUSE to 0. It does the same for PAUSEAUDIO if it's defined (not anymore on latest SDL3?).

So basically what you're saying is that PAUSEAUDIO now follows BLOCK_ON_PAUSE essentially? So background sound rendering still works with this (legacy SDL2) setup code?

**Comment by slouken** [comment]
> So basically what you're saying is that PAUSEAUDIO now follows BLOCK_ON_PAUSE essentially? So background sound rendering still works with this (legacy SDL2) setup code?

Yep!
